### PR TITLE
SCP-2457: new API so `select` can only be applied to contracts that wait

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
`select` picks out the branch that makes progress first. The problem with this is that if one of the branches starts with a non-waiting action, that branch will always be picked. This can be surprising when using `select` with a contract that waits for something but that just happens to start with a non-waiting action.

This PR introduces the type `Promise` which is a `Contract` that starts with a waiting action. `select` now can only be used with `Promise` contracts. `awaitX` actions now have a `Promise` version named `isX`. And `endpoint` now always returns a `Promise`. Bigger `Promise`-type contract can be built using `promiseBind`. To convert a `Promise` back to a `Contract`, use `awaitPromise`.

To rewrite existing code it is usually sufficient to rewrite endpoints as follows:

```haskell
someEndpoint = do
    arg <- endpoint @"some-endpoint"
    ...remainder...
```

becomes

```haskell
someEndpoint = endpoint @"some-endpoint" $ \arg -> do
    ...remainder...
```

and 

```haskell
endpointA `select` endpointB `select` endpointC
```

becomes

```haskell
selectList [endpointA, endpointB, endpointC]
```

and maybe you need to insert `awaitPromise` here and there.

Cases where `select` was used to select between something else than endpoints will need more care.

### CODE REVIEW

File: plutus-contract/src/Plutus/Trace/Effects/RunContract.hs [hunk 1]
```diff
@@ -87,7 +87,7 @@ type ContractConstraints s =
 walletInstanceTag :: Wallet -> ContractInstanceTag
 walletInstanceTag (Wallet i) = fromString $ "Contract instance for wallet " <> show i
 
--- | Run a Plutus contract (client side)
+-- | Run a Plutus contract (client side)activateContractWallet
```
Comment (sorki) [inline plutus-contract/src/Plutus/Trace/Effects/RunContract.hs:90]: Looks like this accidentally slipped in.

**Comment by sorki** [comment]
I wonder if you could update the PR description with short description (i.e. description copied from Jira) - I think I'm struggling with something like this as well (Trace just hanging and eating all my memory after a while) but I'm not sure what does `contracts that wait` really mean - is this a contract waiting for `endpoint` call?

(few minutes later after taking the time to read the PR) - it looks like it's exactly the case as `endpoint` now turns the return type to `Waited` - https://github.com/input-output-hk/plutus/pull/3558/files#diff-c58aeb2f43d03b879e7653c031ff1329b46b0df90277c0286ceaf359678d16dfR339)

I guess this is also related to weird `Game` behavior we are observing with `plutus-starter` in #3526.

p.s.: I find it a bit hard to follow Plutus development as it is quite rapid, there's no changelog and lots of PRs are relevant for just Marlowe - I don't want to create an Issue which makes changelog mandatory and give you even more work at this busy times but I could dedicate some time to writing one which could be helpful for others trying to follow - are there are any plans/timeline for release, tags and changelog and also would you accept a community maintained one if not?

**Comment by j-mueller** [comment]
@sorki : In the `Contract` API we currently have some functions that return immediately (for example `currentSlot` or `balanceTx`), and other functions that wait for a while until an event happens (`awaitSlot`, `endpoint`, `awaitTxConfirmed`, etc). If you mix waiting and non-waiting ones using `select`, then the non-waiting one will always return immediately, which is not very intuitive. So this PR fixes that by changing the types of `select`, `awaitSlot` etc.

> are there are any plans/timeline for release, tags and changelog and also would you accept a community maintained one if not?

Yes we're aware that the changes we're making cause problems for everyone who is using Plutus already. Right now we are focused on getting everything ready for the initial release, but when that is out the door we will adopt a more user-friendly release process with changelogs, tags, etc.

**Comment by sorki** [comment]
Makes sense, thank you @j-mueller !

File: plutus-contract/src/Plutus/Trace/Effects/RunContract.hs (hunk 1 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Trace/Effects/RunContract.hs]: Thanks!

**Review by j-mueller — APPROVED** [review]
This looks very nice now!
